src/atomate2/abinit/files.py

```diff
@@ -0,0 +1,152 @@
+from atomate2.common.files import copy_files, rename_files
+from atomate2.utils.file_client import FileClient, auto_fileclient
+
+__all__ = [
+    "out_to_in",
+    "fname2ext",
+    "load_abinit_input",
+    "load_generator",
+    "write_abinit_input_set",
+]
+
+
+logger = logging.getLogger(__name__)
+
+
+ALL_ABIEXTS = abi_extensions()
+
+
+def fname2ext(filepath):
+    """Get the abinit extension of a given filename.
+
+    This will return None if no extension is found.
+    """
+    filename = os.path.basename(filepath)
+    if "_" not in filename:
+        return None
+    ext = filename.split("_")[-1].replace(".nc", "")
+    if ext not in ALL_ABIEXTS:
+        return None
+    return ext
+
+
+@auto_fileclient
+def out_to_in(
+    out_files: Union[Path, str, dict, Iterable[Union[Path, str, dict]]],
+    src_host: str | None = None,
+    indir: Path | str = INDIR_NAME,
+    file_client: FileClient | None = None,
+    link_files: bool = True,
+):
+    """
+    Copy or link an abinit output file to the Abinit input directory.
+
+    Parameters
+    ----------
+    out_files : str or Path or list
+        The abinit output files to be copied or linked.
+    src_host : str or None
+        The source hostname used to specify a remote filesystem. Can be given as
+        either "username@remote_host" or just "remote_host" in which case the username
+        will be inferred from the current user. If ``None``, the local filesystem will
+        be used as the source.
+    indir : Path or str or None
+        The input directory for Abinit input files.
+    file_client : .FileClient
+        A file client to use for performing file operations.
+    link_files : bool
+        Whether to link the files instead of copying them.
+    """
+    if isinstance(out_files, (Path, str, dict)):
+        out_files = [out_files]
+    for out_file in out_files:
+
+        if isinstance(out_file, dict):
+            if len(out_file) != 1:
+                raise RuntimeError(
+                    "Should have exactly one out file to in file mapping."
+                )
+            out_file, in_file = list(out_file.items())[0]
+            out_file = str(out_file)
+            src_dir = os.path.dirname(out_file)
+            out_file = os.path.basename(out_file)
+        else:
+            out_file = str(out_file)
+            src_dir = os.path.dirname(out_file)
+            out_file = os.path.basename(out_file)
+            in_file = out_file.replace(OUTDATAFILE_PREFIX, INDATAFILE_PREFIX, 1)
+            in_file = os.path.basename(in_file).replace("WFQ", "WFK", 1)
+
+        # Copy or link previous output files to the input directory and rename them
+        copy_files(
+            src_dir=src_dir,
+            dest_dir=indir,
+            src_host=src_host,
+            include_files=[out_file],
+            file_client=file_client,
+            link_files=link_files,
+        )
+        rename_files(
```

Comment:
You are right. Actually, changing the `copy_files` so that it allows to determine the name of the copied file might be complex because that method is also handling file patterns. We could indeed make a copy_file method that works on a file by file basis instead of on a per-directory basis (as what copy_files is intended). Any comments/thoughts about this @utf ?

Reply:
Methods for handling single files already exist on the `FileClient`. Those functions were just wrappers to work with multiple files at once. So maybe this is all you need?

https://github.com/materialsproject/atomate2/blob/5fba05941e9a8faac2f8a13e171e15a43635fbe1/src/atomate2/utils/file_client.py#L220-L240

Reply:
Actually, re-reading the code and this thread, I guess there could be a `copy_and_rename_files` function  that looks very similar to `rename_files` but copies the files instead of moving them.

Reply:
Indeed, I think just using the FileClient methods is fine. I just changed that.